src/coreclr/jit/emitxarch.cpp

```diff
@@ -4153,25 +4153,182 @@ void emitter::emitIns_C(instruction ins, emitAttr attr, CORINFO_FIELD_HANDLE fld
+        case INS_movzx:
+        {
+            return true;
+        }
+
+#if defined(TARGET_AMD64)
+        case INS_movq:
+        case INS_movsxd:
+        {
+            return true;
+        }
+#endif // TARGET_AMD64
+
+        default:
+        {
+            return false;
+        }
+    }
+}
+
+void emitter::emitIns_Mov(instruction ins, emitAttr attr, regNumber dstReg, regNumber srcReg, bool canSkip)
+{
+    // Only move instructions can use emitIns_Mov
+    assert(IsMovInstruction(ins));
+
+#if DEBUG
+    switch (ins)
+    {
+        case INS_mov:
+        case INS_movsx:
+        case INS_movzx:
+        {
+            assert(isGeneralRegister(dstReg) && isGeneralRegister(srcReg));
+            break;
+        }
+
+        case INS_movapd:
+        case INS_movaps:
+        case INS_movdqa:
+        case INS_movdqu:
+        case INS_movsd:
+        case INS_movsdsse2:
+        case INS_movss:
+        case INS_movupd:
+        case INS_movups:
+        {
+            assert(isFloatReg(dstReg) && isFloatReg(srcReg));
+            break;
+        }
+
+        case INS_movd:
+        {
+            assert(isFloatReg(dstReg) != isFloatReg(srcReg));
+            break;
+        }
+
+#if defined(TARGET_AMD64)
+        case INS_movq:
+        {
+            assert(isFloatReg(dstReg) && isFloatReg(srcReg));
+            break;
+        }
+
+        case INS_movsxd:
+        {
+            assert(isGeneralRegister(dstReg) && isGeneralRegister(srcReg));
+            break;
+        }
+#endif // TARGET_AMD64
+
+        default:
+        {
+            unreached();
+        }
+    }
+#endif
+
+    emitAttr size = EA_SIZE(attr);
+
+    assert(size <= EA_32BYTE);
+    noway_assert(emitVerifyEncodable(ins, size, dstReg, srcReg));
+
+    if (canSkip && (dstReg == srcReg))
+    {
+        switch (ins)
+        {
+            case INS_mov:
+            {
+                // These instructions have no side effect and can be skipped
+                return;
+            }
+
+            case INS_movapd:
+            case INS_movaps:
+            case INS_movdqa:
+            case INS_movdqu:
+            case INS_movupd:
+            case INS_movups:
+            {
+                // These instructions have no side effect and can be skipped
+                return;
+            }
+
+            case INS_movd:
+            case INS_movsd:
+            case INS_movsdsse2:
+            case INS_movss:
+            case INS_movsx:
+            case INS_movzx:
+            {
+                // These instructions have a side effect and can't be skipped
+                break;
+            }
+
+#if defined(TARGET_AMD64)
+            case INS_movq:
+            case INS_movsxd:
+            {
+                // These instructions have a side effect and can't be skipped
+                break;
+            }
+#endif // TARGET_AMD64
+
+            default:
+            {
+                unreached();
+            }
+        }
+
+        return;
```

Comment:
Yes, I'll update the comment here.

This is because there are some places on all platforms where the existing code was eliding sign or zero-extensions.

These places should likely be audited since `ins_Copy` (via `inst_Mov`) should be elidable (its just a value copy), but `ins_Move_Extend` (via `inst_Mov_Extend`) should not be, given it may extend.

Reply:
Updated comment to:
```
// These instructions have a side effect and shouldn't be skipped
// however existing codepaths were skipping these instructions in
// certain scenarios and so we skip them as well for back-compat.
//
// Long term, these paths should be audited and should likely be
// replaced with copies rather than extensions.
```

and changed the code to return directly rather than `break` and then `return`